### PR TITLE
Specify a hash for sourceCode

### PR DESCRIPTION
**Pull request by feloy**
This patch gives the possibility to checkout the sources repository to a given commit when displaying source code.

This way, it is possible to display different versions of a file depending on the slide.

```
<h1>Code samples v1</h1>

<split-view>
    <source-code folder="sources"
        hash="cd724e4d106f804fbf02b6dadf3b875a6a17a631"
        files="main.go pod.yaml"
        start-lines="26;18,21"
        end-lines="27;18,22">
    </source-code>
</split-view>

---

<h1>Code samples HEAD</h1>

<split-view>
    <source-code folder="sources"
        files="main.go pod.yaml"
        start-lines="26;18,21"
        end-lines="27;18,22">
    </source-code>
</split-view>


```

### CODE REVIEW

File: handlers/code.go [hunk 1]
```diff
@@ -28,11 +29,23 @@ import (
...
+	"gopkg.in/src-d/go-git.v4/plumbing"
 )
 
 // Code returns the content of a source file.
 func Code(w http.ResponseWriter, r *http.Request) {
 	filename := strings.TrimPrefix(r.URL.Path, "/sourceCode/")
+	hash := r.FormValue("hash")
+
+	if len(hash) > 0 {
+		err := checkout(filepath.Dir(filename), hash)
```
[the remark this answers is not in the document]
Comment (feloy) [reply handlers/code.go]: This new patch is doing a "git clone" of the local repo into memory instead.

[the remark this answers is not in the document]
Comment (feloy) [reply handlers/code.go]: But if you think it's preferable not to copy the entire repo into memory, the "git show" would be possible also:
https://github.com/src-d/go-git/issues/718

Please tell me what you prefer

[the remark this answers is not in the document]
Comment (dgageot) [reply handlers/code.go]: I think git show would be faster. Wouldn’t it?

[the remark this answers is not in the document]
Comment (feloy) [reply handlers/code.go]: Yes, much faster and consuming less memory. I'll use it instead

[the remark this answers is not in the document]
Comment (feloy) [reply handlers/code.go]: I've made the changes

File: handlers/code.go [hunk 2]
```diff
@@ -28,28 +31,85 @@ import (
...
+
+	if len(hash) > 0 {
+
+		repoPath, insidePath, err := findRepositoryOfFile(filename)
+		if err != nil {
+			http.NotFound(w, r)
+			return
+		}
+
+		repo, err := git.PlainOpen(repoPath)
```
Comment (dgageot) [inline handlers/code.go:54]: Can you extract those lines to a function so that you only have to `http.Error` once.

Comment (feloy) [reply handlers/code.go]: done

**Pull request merged**